### PR TITLE
Channel lambda return type inference diagnostics to overload resolution in order to improve error reporting.

### LINKED ISSUE #4587: VB error reporting inside lambdas inside overloads - regressed from VS2013
VB error reporting for some cases has regressed since VS2013. In particular, in VS2013 if there was a overloaded method where all candidates failed, and there was at least one "common" error (same error code and same span) reported by all candidates, then it reported all those common errors, and suppressed other errors, and suppressed the error about "no overload could be found". In VS2015 it doesn't seem to do this any more.

I'm finding this really painful. I've run into it a lot with my async code (especially Task.Run). Also my LINQ code. But here instead is some real-world code I got from a customer...

```vb
Module Module1

    Sub Main()
    End Sub



    Private Async Function GetDataAsync(cs As Characters, imax As Integer) As Task
        Dim Roles = Await cs.GetRoleAsync()

        Dim RoleTasks = Roles.Select(
            Async Function(role As Role) As Task
                Dim Lines = Await role.GetLines()
                If imax <= LinesKey Then Return

                Dim SentenceTasks = Lines.Select(
                    Async Function(Sentence) As Task
                        Dim Words = Await Sentence.GetWordsAsync()
                        If imax <= WordsKey Then Return

                        Dim WordTasks = Words.Select(
                            Async Function(Word) As Task
                                Dim Letters = Await Word.GetLettersAsync()
                                If imax <= LettersKey Then Return

                                Dim StrokeTasks = Letters.Select(
                                    Async Function(Stroke) As Task
                                        Dim endpoints = Await Stroke.GetEndpointsAsync()

                                        Await Task.WhenAll(endpoints.ToArray())
                                    End Function)
                                Await Task.WhenAll(StrokeTasks.ToArray())
                            End Function)
                        Await Task.WhenAll(WordTasks.ToArray())
                    End Function)
                Await Task.WhenAll(SentenceTasks.ToArray())
            End Function)
    End Function



    Function RetryAsync(Of T)(f As Func(Of Task(Of T))) As Task(Of T)
        Return f()
    End Function

End Module


Friend Class Characters
    Function GetRoleAsync() As Task(Of List(Of Role))
        Return Nothing
    End Function
End Class

Class Role
    Function GetLines() As Task(Of List(Of Line))
        Return Nothing
    End Function
End Class

Public Class Line
End Class
```

VS2013 used to give these errors:
 * Error: 'LinesKey' is not declared. It may be inaccessible due to its protection level.
 * Error: 'GetWordsAsync' is not a member of 'ConsoleApplicationVB.Line'.
 * Error: 'WordsKey' is not declared. It may be inaccessible due to its protection level.
 * Error: 'LettersKey' is not declared. It may be inaccessible due to its protection level.

I loved the VS2013 errors because each individual problem was squiggled, so I could do Ctrl+Dot to fix it up.

VS2015 instead just gives this one error:
 * BC30518 Overload resolution failed because no accessible '[Select]' can be called with these arguments:
 * ... Extension method 'Public Function [Select](Of TResult)(selector As Func(Of Role, TResult)) As IEnumerable(Of TResult)' defined in 'Enumerable': Data type(s) of the type parameter(s) cannot be inferred from these arguments. Specifying the data type(s) explicitly might correct this error.
 * ... Extension method 'Public Function [Select](Of TResult)(selector As Func(Of Role, Integer, TResult)) As IEnumerable(Of TResult)' defined in 'Enumerable': Data type(s) of the type parameter(s) cannot be inferred from these arguments. Specifying the data type(s) explicitly might correct this error.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #4587.

@gafter, @jaredpar, @VSadov, @agocke Please review.

### CODE REVIEW

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Invocation.vb [hunk 1]
```diff
@@ -2211,6 +2211,12 @@ ProduceBoundNode:
                             ReportDiagnostic(diagnostics, diagnosticLocation, If(queryMode, ERRID.ERR_TypeInferenceFailureNoExplicitNoBest2, ERRID.ERR_TypeInferenceFailureNoBest2), If(representCandidateInDiagnosticsOpt, candidateSymbol))
                         End If
                     Else
+                        If candidateAnalysisResult.TypeArgumentInferenceDiagnosticsOpt IsNot Nothing AndAlso
```
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Invocation.vb]: If candidateAnalysisResult.TypeArgumentInferenceDiagnosticsOpt?.HasAnyResolvedErrors = True Then

**Comment by gafter** [comment]
LGTM

**Comment by AlekseyTs** [comment]
Ping. @VSadov, @jaredpar, @agocke Please review. Need a second sign-off for this change.

**Comment by agocke** [comment]
:+1:

**Pull request merged**